### ISSUE TITLE
Scaling factor config.loss_s in ArcFace loss: two conflicting implementations

### ISSUE THREAD

**Issue by jolastar** [deepinsight/insightface#1011]
In the ArcFace article, the _s_ scaling factor is applied to all the logits after the margin has been added on the logit that corresponds to the target class.

In the code here, `recognition/train_parall.py` seems to conform to the description in the article but I see an unexplained variation in `recognition/train.py`... Is there an explanation for the variant in `recognition/train.py`?

This version seems consistent with the article (the `config.loss_s` factor is applied at the end on all the logits):

https://github.com/deepinsight/insightface/blob/b1c747e4d58dba0dd011df4cd813ff70d828449c/recognition/train_parall.py#L100-L114

But this version is weird and unexplained - the logit of the target class is _divided_ by `config.loss_s`, then the margin is applied, then the modified logit is multiplied by `config.loss_s`, while the other logits are not affected by any scaling:

https://github.com/deepinsight/insightface/blob/4a4b8d03fec981912fdef5b3232a37a827cbeed6/recognition/train.py#L91-L105

**Comment by TMaysGGS** [comment]
I think this should be the implementation of ArcFace loss and it matches other implementations like in TensorFlow. 
`    s = args.margin_s
    m = args.margin_m
    assert s>0.0
    assert m>=0.0
    assert m<(math.pi/2)
    _weight = mx.symbol.L2Normalization(_weight, mode='instance')
    nembedding = mx.symbol.L2Normalization(embedding, mode='instance', name='fc1n')*s
    fc7 = mx.sym.FullyConnected(data=nembedding, weight = _weight, no_bias = True, num_hidden=args.num_classes, name='fc7')
    zy = mx.sym.pick(fc7, gt_label, axis=1)
    cos_t = zy/s
    cos_m = math.cos(m)
    sin_m = math.sin(m)
    mm = math.sin(math.pi-m)*m
    #threshold = 0.0
    threshold = math.cos(math.pi-m)
    if args.easy_margin:
      cond = mx.symbol.Activation(data=cos_t, act_type='relu')
    else:
      cond_v = cos_t - threshold
      cond = mx.symbol.Activation(data=cond_v, act_type='relu')
    body = cos_t*cos_t
    body = 1.0-body
    sin_t = mx.sym.sqrt(body)
    new_zy = cos_t*cos_m
    b = sin_t*sin_m
    new_zy = new_zy - b
    new_zy = new_zy*s
    if args.easy_margin:
      zy_keep = zy
    else:
      zy_keep = zy - s*mm
    new_zy = mx.sym.where(cond, new_zy, zy_keep)

    diff = new_zy - zy
    diff = mx.sym.expand_dims(diff, 1)
    gt_one_hot = mx.sym.one_hot(gt_label, depth = args.num_classes, on_value = 1.0, off_value = 0.0)
    body = mx.sym.broadcast_mul(gt_one_hot, diff)
    fc7 = fc7+body`

**Comment by jolastar** [comment]
My bad, sorry: I overlooked a line in [recognition/train.py](https://github.com/deepinsight/insightface/blob/master/recognition/train.py) where the output of `fc1` is multiplied by `config.loss_s` after normalization, before being fed to `fc7`. So effectively, the logits are indeed all scaled by `config.loss_s`, like in [recognition/train_parall.py](https://github.com/deepinsight/insightface/blob/master/recognition/train_parall.py).

The `*s` at the end of the line was easy to miss:
https://github.com/deepinsight/insightface/blob/4a4b8d03fec981912fdef5b3232a37a827cbeed6/recognition/train.py#L82

(@TMaysGGS: indeed I see the implementation at [src/train_softmax.py](https://github.com/deepinsight/insightface/blob/master/src/train_softmax.py), and I see that it's been used as basis for TF implementation, but it's more complicated and in principle the `src` folder is superseded by `recognition`. So I prefer to use the straightforward implementation at [recognition/train_parall.py](https://github.com/deepinsight/insightface/blob/master/recognition/train_parall.py).)
